### ISSUE TITLE
Scan for vulnerable NodeJS packages via Travis

### ISSUE THREAD

**Issue by chillu** [silverstripe/silverstripe-admin#764]
There's been a couple of high profile compromises of downstream dependencies in the NodeJS ecosystem. We should build security checks for this into our CI process, and fail the build to surface this early. See https://blog.risingstack.com/controlling-node-js-security-risk-npm-dependencies/

**Comment by blueo** [comment]
this is possibly an option: https://snyk.io/ can scan php dependencies too.

**Comment by chillu** [comment]
Synk runs their own [PHP Vulnerability Database](https://snyk.io/docs/snyk-for-php) so we'd need to report issues in our own modules there, as well as encouraging the community to do so. The [Symfony Security](https://github.com/FriendsOfPHP/security-advisories) project seems more likely to get traction? 

Symfony costs money (at least 2EUR per month), which is generally fine in terms of supporting friendly open source colleagues. But the "per project" bit is a bit worrying: We have a LOT of projects to check (70+ supported modules). I've contacted them with the following message:

--- 

Hello! 

We're running an open source PHP CMS (silverstripe.org) with dozens of modules supported by a dedicated team (https://www.silverstripe.org/software/addons/silverstripe-commercially-supported-module-list/). Incidentally we're using a lot of Symfony components, so thanks for creating awesome software!

We'd like to tighten up our security practices around secure dependencies, so looking to integrate Symfony Security through API checks in our Travis-based CI. We're generally happy to pay a small amount towards this, since we're running a business on top of these open source offerings.

The tricky part is what you consider a "project". We have quite a lot of modules, each with their own CI builds. Do you interpret "project" as "repo"? In which case your service would likely be out of our price range. Or a more loose definition of "project" where multiple modules could upload differing composer.lock files?

Thanks, and all the best from Wellington, New Zealand
Ingo

**Comment by maxime-rainville** [comment]
Silly suggestion, but GitHub seems to be pretty good at notifying us when we're using vulnerable node dependencies. Is there a problem with relying on that?

The PHP vulnerability side of things is indirectly addressed by https://github.com/silverstripe/silverstripe-framework/issues/8548

**Comment by chillu** [comment]
Github only covers NodeJS dependencies for us, it doesn't scan PHP. Ideally we have one solution which does both. Even if it did scan PHP, we've got repos outside of Github (e.g. in CWP Gitlab). I'm aiming for the least amount of solutions covering the most surface here. Might not be practical though.

FYI, I haven't heard back from Symfony Security.

**Comment by chillu** [comment]
Of course I missed one: https://dependabot.com/ has been acquired by Github, and does this all for free. Including PHP! I can't see where it gets security advisories from in the case of PHP (since they're not provided by composer/packagist directly)

**Comment by chillu** [comment]
It appears that the "auto-merge" feature is optional (as it should be), but Dependabot always asks for write permissions to code in the repo. That's not cool, I don't want it to have that level of access.

![image](https://user-images.githubusercontent.com/111025/61916724-b69d5000-af9d-11e9-86ed-77c1d76a598a.png)

I've contacted them about the possibility to set more granular permissions

**Comment by chillu** [comment]
Answer from Dependabot:

![image](https://user-images.githubusercontent.com/111025/63309884-6ae78780-c34c-11e9-8913-33ca3023aee7.png)

**Comment by chillu** [comment]
Github has Composer/PHP support for this now! https://github.blog/2019-09-18-dependency-graph-supports-php-repos-with-composer-dependencies/

**Comment by brynwhyman** [comment]
Just a note on Dependabot - it creates branches on our repositories, which will show up in places like Packagist. Not the end of the world, but it's something we currently look to avoid with things like community pull requests (which should come from a fork).

**Comment by brynwhyman** [comment]
Noting that it looks like GitHub is rolling out Dependabot to all repositories by default anyway, we're seeing them pop up on our commercially supported modules already.

Given this is being rolled out more widely, having Dependabot creating branches on origin might not be that big of a deal? After a merge they're deleted automatically anyway.

See https://help.github.com/en/github/managing-security-vulnerabilities/configuring-automated-security-fixes:
> We'll automatically enable automated security fixes in every repository that uses security alerts and the dependency graph over the next few months, starting in May 2019.

**Comment by ScopeyNZ** [comment]
You can still opt out. Annoying that they're enabling by default though. Most cases they can be merged without worrying, but in some cases dist files need to be rebuilt (implying that the security issue actually matters - it doesn't if dist files don't need changing) meaning that the PR is a bit useless.

In summary, they are pointless and can be ignored if the dist files aren't affected. They should be done manually if they require rebuilt dist files. I still think we should opt out for now.

**Comment by chillu** [comment]
> In summary, they are pointless and can be ignored if the dist files aren't affected. They should be done manually if they require rebuilt dist files. I still think we should opt out for now.

Agreed. The important part is the triage ("does this issue affect our users?"), and tracking issues. I think unless we deem it a high/critical dependency issue, we can just deal with this by a regular sweep across the repos. That seems more viable than scattering everyone's attention by checking out PRs, rebuilding dist, pushing, waiting for build to pass, merging - every time a dependency issue comes up.

Maybe there's a way that we could auto-fix those dependabot PRs through Github Actions or something? Might be more hassle than it's worth for now to build this of course.

**Comment by robbieaverill** [comment]
I also wondered whether you can opt out of this new feature in bulk across the organisation - that'd be nice

**Comment by chillu** [comment]
Argh, I've just added one criteria which throws off our plans to use Github's built in advisories:

>  * Can inspect all actively maintained release branches (not just the default branch)

According to https://github.community/t5/How-to-use-Git-and-GitHub/Is-it-possible-to-get-security-updates-for-a-branch-other-than/m-p/21676, that's not possible. So in the case of `silverstripe/framework`, we'd get alerts for 4.x, but not 3.x.

**Comment by chillu** [comment]
Also, it looks like just because a vulnerability has been reported as a CVE, even though Github claims to use that as a source, there's cases where it doesn't come through (see [forum post](https://github.community/t5/How-to-use-Git-and-GitHub/Security-Alerts/td-p/26425)). That's a Java/POM dependency, maybe it's more clear cut with PHP/Composer.

**Comment by chillu** [comment]
Update on this: Snyk has greatly improved their PHP support, see ["silverstripe" search](https://snyk.io/vuln/search/page/1?type=composer&q=silverstripe), [php instructions](https://support.snyk.io/hc/en-us/articles/360003817397-Snyk-for-PHP) and [silverstripe/framework overview](https://snyk.io/vuln/composer:silverstripe%2Fframework)

**Comment by chillu** [comment]
Update from Github support:

> We've removed FriendsOfPHP as a listed source for GitHub Advisory Database until we're able to do a full import of FriendsOfPHP data.
